### ISSUE TITLE
Attach detach controller not working in k8s 1.3 for AWS EBS volume

### ISSUE THREAD

**Issue by dhawal55** [kubernetes/kubernetes#28643]
I upgraded to the latest k8s 1.3 and trying to leverage the latest attach detach controller. But I'm unable to mount AWS EBS volumes. I get the below error:

Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "mypod-2170121488-bh7ir"/"kube-system". list of unattached/unmounted volumes=[my-storage]

As per AWS, the volume status is in-use and attached but k8s is unable to mount it correctly on the pod as specified in the error above.

**Comment by dhawal55** [comment]
Doing some troubleshooting and found the following error messages in kubelet logs:

```
Jul 08 16:31:32 ip-a-b-c-d.cloud.nordstrom.net kubelet[13953]: I0708 16:31:32.862792   13953 reconciler.go:214] AttachVolume operation started for volume "kubernetes.io/aws-ebs/vol-795ffbf0" (spec.Name: "influxdb-storage") pod "82b21a9d-449a-11e6-bdee-028831ba1e41" (UID: "82b21a9d-449a-11e6-bdee-028831ba1e41")
Jul 08 16:31:33 ip-a-b-c-d.cloud.nordstrom.net kubelet[13953]: E0708 16:31:33.757487   13953 attacher.go:66] Error attaching volume "vol-795ffbf0": error finding instance ip-a-b-c-d.cloud.nordstrom.net: no instances found for name: ip-a-b-c-d.cloud.nordstrom.net
Jul 08 16:31:33 ip-a-b-c-d.cloud.nordstrom.net kubelet[13953]: E0708 16:31:33.757524   13953 goroutinemap.go:155] Operation for "kubernetes.io/aws-ebs/vol-795ffbf0" failed. No retries permitted until 2016-07-08 16:33:33.75751862 +0000 UTC (durationBeforeRetry 2m0s). error: AttachVolume.Attach failed for volume "kubernetes.io/aws-ebs/vol-795ffbf0" (spec.Name: "influxdb-storage") from node "ip-a-b-c-d.cloud.nordstrom.net" with: error finding instance ip-a-b-c-d.cloud.nordstrom.net: no instances found for name: ip-a-b-c-d.cloud.nordstrom.net
```
This looks like a similar issue to one we had earlier this year where AWS provider was querying for instance by PrivateDNS name instead of instance ID and not able to find an instance (if your VPC uses custom domain name like ours).

**Comment by dhawal55** [comment]
The AttachDisk and DetachDisk methods are passing in the hostname to getAwsInstance method (https://github.com/kubernetes/kubernetes/blob/f2ddd60eb9e7e9e29f7a105a9a8fa020042e8e52/pkg/cloudprovider/providers/aws/aws.go#L1317). Any idea why the getAwsInstance method still supports nodeName? It keeps causing issues like this. It may be time to update the method to not take any parameters.

**Comment by justinsb** [comment]
@dhawal55 let's figure this out once and for all :-)

The node name needs to be resolvable from the master.  Can you please tell me:

1. The PrivateDNSName as reported in one of your nodes in the API
2. The instance name from the AWS metadata (local-hostname and public-hostname)
3. What the node name is in the K8s API
4. Your VPC domain name
5. How you make sure that the instance names are resolvable

As I recall the root problem is that AWS returns different values for 1 and 2.  See https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/aws/aws.go#L1301-L1308

**Comment by dhawal55** [comment]
@justinsb Won't it be just easier to not pass nodeName to getAwsInstance method and always get the instance from selfAWSInstance (which uses instanceID to query the instance). I have created a pull request with the changes - https://github.com/kubernetes/kubernetes/pull/28724. Please let me know if it makes sense.

To answer your above questions:
 1. PrivateDNSName in AWS: ip-a-b-c-d.us-west-2.compute.internal
 2. local-hostname: ip-a-b-c-d.cloud.nordstrom.net; hostname: ip-a-b-c-d.cloud.nordstrom.net
 3. node name in k8s: ip-a-b-c-d.us-west-2.compute.internal
 4. VPC domain name: cloud.nordstrom.net
 5. This is the same issue we faced earlier (https://github.com/kubernetes/kubernetes/issues/11543). When you use a custom domain name in the VPC, the hostname in metadata represents the custom domain but when your query AWS ec2 PrivateDNSName by hostname, you don't get any results as the PrivateDNSName uses us-west-2.compute.internal domain. What is interesting in this case is the node name in k8s is correct (uses the AWS default domain), so if the AttachDisk code looks-up using the k8s node name instead of hostname, we will probably not have any issue.

**Comment by dhawal55** [comment]
@justinsb Any update on this? Looks like there are lots of issues related to volume mounting.

**Comment by justinsb** [comment]
Sorry for the delay @dhawal55 .

Looks like a confusion between hostname & nodename in the volume code.  I'm currently looking at going through the whole codebase and replacing every `nodeName string` with `nodeName types.NodeName`.... wish me luck!

Have found the bug I think you're hitting, but may still find others...

**Comment by dhawal55** [comment]
@justinsb Thanks for the update. Just wanted to check the status. Let me know if i can help in any way.

**Comment by justinsb** [comment]
Fix is (buried) in #29048.  I would obviously prefer to fix it right, but if it's problematic we can split out the actual fix I guess!

**Comment by dhawal55** [comment]
@justinsb Thanks for the pull request. I like the change but i'm still not clear how this will fix the issue. From what i understand, the change is mostly replacing the type for nodeName from `string` to `types.NodeName`.

We are assigning the hostname to nodeName right here [https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L636]. In my case, hostname (uname -n) is `ip-a-b-c-d.cloud.nordstrom.net` so nodeName will also be set to the same value and node lookup will still fail.

**Comment by dhawal55** [comment]
@justinsb I think i found the root cause for the error. When the kubelet (on worker node) creates the volume manager, it passes the hostName to it instead of the nodeName [https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L525]. The Volume manager when attempting to mount the volume, calls the API server to lookup the node (using hostName instead of nodeName) to verify that the controller has attached the volume and gets the below error:
`goroutinemap.go:155] Operation for "kubernetes.io/aws-ebs/vol-f310917a" failed. No retries permitted until 2016-07-19 18:15:41.073705591 +0000 UTC (durationBeforeRetry 2m0s). error: VerifyControllerAttachedVolume failed fetching node from API server. Volume "kubernetes.io/aws-ebs/vol-f310917a" (spec.Name: "influxdb-storage") pod "b5419798-4dd9-11e6-8273-026731bc9baf" (UID: "b5419798-4dd9-11e6-8273-026731bc9baf"). Error: nodes "ip-172-24-134-153.cloud.nordstrom.net" not found.`

The error i initially posted didn't had this message probably because while troubleshooting i might have run kubelet with `--enable-controller-attach-detach=false`.

I will update kubelet code to pass nodeName to `NewVolumeManager` and test it. If it works, i will submit a pull request. Thanks for all your help.

**Comment by mikekap** [comment]
On a sort-of related note, I noticed that `volumeManager` is initialized twice:
 - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L409
 - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L517

That doesn't look intentional, though I'm not very familiar with the code.

**Comment by w0mbat** [comment]
I also tried this one [here](https://github.com/kubernetes/kubernetes/blob/release-1.2/examples/mysql-cinder-pd/mysql.yaml) on Rackspace with a clean K8s 1.3 and also get the same error (of course I changed the volumeID) and also added the correct Rackspace credentials in the cloud-config

**Comment by justinsb** [comment]
@mikekap I agree that doesn't look right - can you open a separate issue for it please (because I think we'll close this one soon!)

**Comment by justinsb** [comment]
@dhawal55 that is the fix, yes: https://github.com/kubernetes/kubernetes/pull/29048/files#diff-bf28da68f62a8df6e99e447c4351122dL507

If you want to do a PR with just that fix, please go ahead - it should get it faster & will be easier to cherry-pick.  If you CC me I can LGTM it.

**Comment by dhawal55** [comment]
@justinsb I see that along with type changes, you have actually fixed `NewVolumeManager` initialization to pass nodeName. I tested the fix and it works. Here's the pull request: https://github.com/kubernetes/kubernetes/pull/29240

How do i merge this fix for v1.3.3-beta.0? I don't see a branch for 1.3.3.

**Comment by justinsb** [comment]
There's a 1.3 branch, once the PR is approved you can propose it for a cherry-pick: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/cherry-picks.md

Periodically (whenever there's something important enough on the 1.3 branch) a revision tag & release is made (i.e. 1.3.3)

**Comment by dpratt** [comment]
I can't seem to find a cherry pick of this into the recently released 1.3.3, but it's entirely possible that I've missed it. Is this going to be in any of the 1.3.x builds, or do I have to roll my own off master until 1.4 comes out?

**Comment by dhawal55** [comment]
I just created a pull request to merge this fix into release-1.3 branch https://github.com/kubernetes/kubernetes/pull/29554. Hopefully, we will get this fix in v1.3.4

**Comment by massimiliano-mantione** [comment]
I have the *same* problem with block storage volumes on GCE (I mean the GCE
equivalent of EBS).
Will the fix also fix the issue on GCE?

**Comment by richarvey** [comment]
Has this made it into v1.3.4? because we are having this issue on 1.3.5 on coreOS

**Comment by sekka1** [comment]
Same here, just ran into this issue.  Any updates on this?
```
Server Version: version.Info{Major:"1", Minor:"3", GitVersion:"v1.3.5", GitCommit:"b0deb2eb8f4037421077f77cb163dbb4c0a2a9f5", GitTreeState:"clean", BuildDate:"2016-08-11T20:21:58Z", GoVersion:"go1.6.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by justinsb** [comment]
Fix was in 1.3.4.

Also @massimiliano-mantione this is AWS only - wouldn't affect GCE unless you're doing something really odd!

I suggest opening separate issues if you are seeing this on 1.3.4 or 1.3.5
